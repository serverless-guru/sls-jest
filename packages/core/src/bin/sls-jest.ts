#!/usr/bin/env node
import { program } from 'commander';
import fs from 'fs';
import path from 'path';
import { destroyAllStacks } from '../infrastructure';

const pathName = fs.realpathSync(path.join(__dirname, '../../package.json'));
const { version } = JSON.parse(fs.readFileSync(pathName, 'utf8'));

program
  .name('sls-jest')
  .version(version, '-v, --version', 'outputs the current version')
  .description('SLS Jest CLI');

program
  .command('destroy')
  .requiredOption('-t, --tag <tag>', 'Tag to use for the stack')
  .action(async (args: { tag: string }) => {
    await destroyAllStacks({
      tag: args.tag,
    });
  });

program.parseAsync();
