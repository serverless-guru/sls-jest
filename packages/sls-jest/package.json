{
  "name": "@sls-jest/core",
  "version": "0.0.0",
  "description": "A collection of matchers and spies for testing AWS serverless applications",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "files": [
    "lib"
  ],
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1",
    "build": "npx rimraf ./lib && tsc"
  },
  "author": "Serverless Guru LLC",
  "license": "ISC",
  "devDependencies": {
    "@sls-jest/infrastructure": "*",
    "@tsconfig/node16": "^1.0.3",
    "@types/aws-lambda": "^8.10.101",
    "@types/jest": "^28.1.6",
    "@types/lodash": "^4.14.182",
    "@types/node": "^18.6.3",
    "aws-cdk": "^2.35.0",
    "aws-lambda": "^1.0.7",
    "prettier": "^2.7.1",
    "rimraf": "^3.0.2",
    "ts-node": "^10.9.1",
    "ts-toolbelt": "^9.6.0"
  },
  "dependencies": {
    "@aws-sdk/client-appsync": "^3.142.0",
    "@aws-sdk/client-cloudwatch-logs": "^3.137.0",
    "@aws-sdk/client-dynamodb": "^3.137.0",
    "@aws-sdk/client-sqs": "^3.142.0",
    "@aws-sdk/lib-dynamodb": "^3.145.0",
    "@jest/expect-utils": "^28.1.3",
    "aws-cdk-lib": "^2.35.0",
    "constructs": "^10.1.67",
    "jest-matcher-utils": "^28.1.3",
    "jest-snapshot": "^28.1.3",
    "json-canonicalize": "^1.0.4",
    "lodash": "^4.17.21",
    "rxjs": "^7.5.6",
    "typescript": "^4.7.4"
  }
}
