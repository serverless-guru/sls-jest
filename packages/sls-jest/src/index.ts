import { EventBridgeEvent } from 'aws-lambda';
import { DocumentClient } from 'aws-sdk/clients/dynamodb';

export * from './spies/eventBridge';
export * from './matchers/helpers';
export * from './EventBridge';

declare global {
  namespace jest {
    interface Matchers<R> {
      toHaveItem(key: DocumentClient.Key): R;
      toHaveItems(key: DocumentClient.Key[]): R;
      toHaveEventMatchingObject(
        expected: Partial<EventBridgeEvent<string, unknown>>,
      ): R;
      toHaveEventMatchingObjectTimes(
        expected: Partial<EventBridgeEvent<string, unknown>>,
        times: number,
      ): R;
      toEvaluateTo(template: string | object): R;
      toEvaluateToSnapshot(propertiesOrHint?: string, hint?: string): R;
      toEvaluateToInlineSnapshot(propertiesOrHint?: string, hint?: string): R;
    }
  }
}
