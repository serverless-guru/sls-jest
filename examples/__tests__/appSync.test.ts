import { appSyncMappingTemplate } from 'sls-jest';

type DynamoDBGetItem = {
  version: string;
  operation: string;
  key: {
    pk: { S: string };
  };
};

const template = `
#set($id=$ctx.args.id)
{
  "version" : "2017-02-28",
  "operation" : "GetItem",
  "key" : {
      "pk" : $util.dynamodb.toDynamoDBJson($id)
  }
}
`;

describe('Mapping Template', () => {
  it('should evaluate a template string', async () => {
    const expected = `
{
  "version" : "2017-02-28",
  "operation" : "GetItem",
  "key" : {
      "pk" : {"S":"123"}
  }
}
`;

    // test that the template evaluates to the expected value as a string
    await expect(
      appSyncMappingTemplate({
        template,
        context: {
          arguments: {
            id: '123',
          },
        },
      }),
    ).toEvaluateTo(expected);
  });

  it('should evaluate a template object', async () => {
    // test that the template evaluates to the expected value as an object
    await expect(
      appSyncMappingTemplate({
        template,
        context: {
          arguments: {
            id: '123',
          },
        },
      }),
    ).toEvaluateTo<DynamoDBGetItem>({
      version: '2017-02-28',
      operation: 'GetItem',
      key: {
        pk: { S: '123' },
      },
    });
  });

  it('should evaluate a template snapshot as object', async () => {
    // test that the template evaluates to the expected snapshot
    // if the snapshot evaluates to an object, it is parsed before being saved
    await expect(
      appSyncMappingTemplate({
        template,
        context: {
          arguments: {
            id: '456',
          },
        },
      }),
    ).toEvaluateToSnapshot();
  });

  it('should evaluate a template inline snapshot as object', async () => {
    // test that the template evaluates to the expected inline snapshot
    // if the snapshot evaluates to an object, it is parsed before being saved
    await expect(
      appSyncMappingTemplate({
        template,
        context: {
          arguments: {
            id: '789',
          },
        },
      }),
    ).toEvaluateToInlineSnapshot(`
      Object {
        "key": Object {
          "pk": Object {
            "S": "789",
          },
        },
        "operation": "GetItem",
        "version": "2017-02-28",
      }
    `);
  });

  it('should evaluate a template snapshot as string', async () => {
    // test that the template evaluates to the expected snapshot
    await expect(
      appSyncMappingTemplate({
        template: 'hello ${ctx.args.id}',
        context: {
          arguments: {
            id: '456',
          },
        },
      }),
    ).toEvaluateToSnapshot();
  });

  it('should evaluate a template inline snapshot', async () => {
    // test that the template evaluates to the expected inline snapshot
    await expect(
      appSyncMappingTemplate({
        template: 'hello ${ctx.args.id}',
        context: {
          arguments: {
            id: '789',
          },
        },
      }),
    ).toEvaluateToInlineSnapshot(`"hello 789"`);
  });
});
