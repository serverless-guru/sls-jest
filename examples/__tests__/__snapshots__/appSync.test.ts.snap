// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JS resolvers should evaluate a js resolver with snapshot 1`] = `
Object {
  "key": Object {
    "id": Object {
      "S": "123",
    },
  },
  "operation": "GetItem",
}
`;

exports[`Mapping Template should evaluate a template snapshot 1`] = `
Object {
  "key": Object {
    "pk": Object {
      "S": "456",
    },
  },
  "operation": "GetItem",
  "version": "2017-02-28",
}
`;
